artichoke-backend/src/extn/core/time/offset.rs

```diff
@@ -0,0 +1,211 @@
+//! Parser for Ruby Time offset parameter to help generate `Time`.
+
+use crate::convert::{implicitly_convert_to_int, implicitly_convert_to_string};
+use crate::extn::core::symbol::Symbol;
+use crate::extn::core::time::Offset;
+use crate::extn::prelude::*;
+
+#[allow(clippy::cast_precision_loss)]
+const MAX_FLOAT_OFFSET: f64 = i32::MAX as f64;
+#[allow(clippy::cast_precision_loss)]
+const MIN_FLOAT_OFFSET: f64 = i32::MIN as f64;
+
+impl TryConvertMut<Value, Option<Offset>> for Artichoke {
+    type Error = Error;
+
+    fn try_convert_mut(&mut self, options: Value) -> Result<Option<Offset>, Self::Error> {
+        let hash: Vec<(Value, Value)> = self.try_convert_mut(options)?;
+
+        // Short circuit. An empty options hash does not error.
+        //
+        // Example:
+        //
+        // ```console
+        // [2.6.3]> Time.at(0, {})
+        // => 1970-01-01 01:00:00 +0100
+        // ```
+        if hash.is_empty() {
+            return Ok(None);
+        }
+
+        // All other keys are rejected.
+        //
+        // Example:
+        // ```console
+        // [2.6.3]> Time.at(0, i: 0)
+        // ArgumentError (unknown keyword: i)
+        // ```
+        for (mut key, _) in &hash {
```

Comment:
Agreed per the conversation we had on Discord. This looks like we should extract this to the `convert` module.

`String::new` wants the same functionality with its `capacity:` kwarg:

```
[3.1.2] > String.new(capacity: 10)
=> ""
[3.1.2] > String.new(capa: 10)
(irb):4:in `initialize': unknown keyword: :capa (ArgumentError)
        from (irb):4:in `new'
        from (irb):4:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
[3.1.2] > String.new(capa: 10)
^C
[3.1.2] > String.new(capacity: 10, nine: 10)
(irb):5:in `initialize': unknown keyword: :nine (ArgumentError)
        from (irb):5:in `new'
        from (irb):5:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
```

Reply:
Filed:

- https://github.com/artichoke/artichoke/issues/2091